### ISSUE TITLE
Azurerm_cosmosdb_account missing ID in location when updating resource, so the azure is trying to recreate the locations

### ISSUE THREAD

**Issue by rfojta** [hashicorp/terraform-provider-azurerm#8730]
### Description

When modifying comsosdb account on any other properties than locations, Azure understand that it should create locations because they're send to the API without IDs.

Confirmed by email communication with MS Support - Yazan Flaifel <Yazan.Flaifel@microsoft.com>

Quote:

The issue here is that you have non-standard regional account Ids. The current regional account Ids are 
"abc-123-rep-0" and "abc-123-rep-1" respectively for East US 2 and Central US regions.

Due to these non-standard Ids, when you call PUT again, you must explicitly set these non-standard Ids. It should be something like this:
```
{
  "locations" : [
    {
      "id": "abc-123-rep-1",
      "locationName":"centralus",
      "failoverPriority":1
    },
    {

      "id": "abc-123-rep-0",
      "locationName":"eastus2",
      "failoverPriority":0
    }
  ]
}
```
If the non-standard id is not specified, then we default to <databaseAccountName>-<region>format. 
This causes a mismatch with the existing regional database account Ids, causing the validation error, which is the error you observed.

Regards,
Yazan Flaifel
ARR SUPPORT ENG
Azure Rapid Response
Office: +962 (6) 5564137
yazan.flaifel@microsoft.com 
10 AM – 7 PM (UTC +3)

End of quote

### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Terraform (and AzureRM Provider) Version

2.xx

### Affected Resource(s)

* `azurerm_cosmosdb_account`

### Terraform Configuration Files

```hcl
resource "azurerm_cosmosdb_account" "mongodb_account" {
  name                      = var.resource_tags.name
  location                  = var.region
  resource_group_name       = var.resource_group_name
  offer_type                = var.offer_type
  kind                      = "MongoDB"
  tags                      = merge(module.base_tags.citrix_tags_map.triggers, var.resource_tags)
  enable_automatic_failover = var.enable_automatic_failover
  enable_multiple_write_locations = var.enable_multiple_write_locations

  dynamic capabilities { 
    for_each = var.capabilities
    content {
      name = capabilities.value
    }
  }

  consistency_policy {
    consistency_level       = var.consistency_level
    max_interval_in_seconds = var.max_interval_in_seconds
    max_staleness_prefix    = var.max_staleness_prefix
  }

  dynamic "geo_location" {
    for_each = split(",", format("%s,%s", var.region, join(",", var.failover_regions)))

    content {
      location          = geo_location.value
      failover_priority = geo_location.key
      prefix            = format("%s-%d", var.prefix, geo_location.key)
    }
  }
  is_virtual_network_filter_enabled = var.is_virtual_network_filter_enabled
  ip_range_filter = var.ip_range_filter
}
```

### Expected Behavior

PUT /subscriptions/abcdefg-123456/resourceGroups/abc-123-rg/providers/Microsoft.DocumentDB/databaseAccounts/abc-123?api-version=2020-04-01 HTTP/1.1
```js
{
    "kind": "MongoDB",
    "location": "southeastasia",
    "properties": {
        "consistencyPolicy": {
            "defaultConsistencyLevel": "Session",
            "maxStalenessPrefix": 100,
            "maxIntervalInSeconds": 5
        },
        "locations": [
            {
                "id": "abc-123-rep-0",
                "failoverPriority": 0,
                "locationName": "Southeast Asia"
            },
            {
                "id": "abc-123-rep-1",
                "failoverPriority": 1,
                "locationName": "East Asia"
            }
        ],
        "databaseAccountOfferType": "Standard",
        "ipRules": [
            {
                "ipAddressOrRange": "xx.xx.xx.xx"
            },
            {
                "ipAddressOrRange": "yy.yy.yy.yy"
            }
        ],
        "isVirtualNetworkFilterEnabled": false,
        "enableAutomaticFailover": true,
        "capabilities": [
            {
                "name": "EnableMongo"
            }
        ],
        "virtualNetworkRules": [],
        "enableMultipleWriteLocations": true,
        "enableFreeTier": false
    },
    "tags": {
    }
}
```

### Actual Behavior

PUT /subscriptions/abcdefg-123456/resourceGroups/abc-123-rg/providers/Microsoft.DocumentDB/databaseAccounts/abc-123?api-version=2020-04-01 HTTP/1.1

```js
{
    "kind": "MongoDB",
    "location": "southeastasia",
    "properties": {
        "consistencyPolicy": {
            "defaultConsistencyLevel": "Session",
            "maxStalenessPrefix": 100,
            "maxIntervalInSeconds": 5
        },
        "locations": [
            {
                "failoverPriority": 0,
                "locationName": "Southeast Asia"
            },
            {
                "failoverPriority": 1,
                "locationName": "East Asia"
            }
        ],
        "databaseAccountOfferType": "Standard",
        "ipRules": [
            {
                "ipAddressOrRange": "xx.xx.xx.xx"
            },
            {
                "ipAddressOrRange": "yy.yy.yy.yy"
            }
        ],
        "isVirtualNetworkFilterEnabled": false,
        "enableAutomaticFailover": true,
        "capabilities": [
            {
                "name": "EnableMongo"
            }
        ],
        "virtualNetworkRules": [],
        "enableMultipleWriteLocations": true,
        "enableFreeTier": false
    },
    "tags": {
    }
}
```

Error: Error updating CosmosDB Account "abc-123" properties (Resource Group "abc-123-rg"): Error creating/updating CosmosDB Account "iws-stage-seas-uns-mongo36" (Resource Group "iws-stage-seas-uns"): documentdb.DatabaseAccountsClient#CreateOrUpdate: Failure sending request: StatusCode=400 -- Original Error: Code="BadRequest" Message="Regions Southeast Asia already exists for database account abc-123\r\nActivityId: cd75c776-cc8e-4909-b671-afb7f3dc43bc, Microsoft.Azure.Documents.Common/2.11.0"

### Steps to Reproduce

1. `terraform apply`

### References

Here's the code that needs fixing by adding ID

* Method expandAzureRmCosmosDBAccountGeoLocations: https://github.com/terraform-providers/terraform-provider-azurerm/blob/master/azurerm/internal/services/cosmos/cosmosdb_account_resource.go#L842

* #0000

**Comment by rfojta** [comment]
Correction 
If the non-standard id is not specified, then we default to`<databaseAccountName>-<region>format. `

**Comment by yupwei68** [comment]
Hi @rfojta , thanks for opening this issue. Sorry that I could not reproduce this error if I understand the steps correctly. 
First I have applied a cosmos account as bellow:
```
provider "azurerm" {
  features {}
}

resource "azurerm_resource_group" "test" {
  name     = "test-cos-yup"
  location = "West Europe"
}

resource "azurerm_cosmosdb_account" "mongodb_account" {
  name                              = "yup-cos-geo"
  location                          = azurerm_resource_group.test.location
  resource_group_name               = azurerm_resource_group.test.name
  offer_type                        = "Standard"
  kind                              = "MongoDB"
  enable_automatic_failover         = true
  is_virtual_network_filter_enabled = false

  capabilities {
    name = "EnableMongo"
  }

  consistency_policy {
    consistency_level = "Session"
  }

  geo_location {
    location          = "Southeast Asia"
    failover_priority = 0
  }

  geo_location {
    location          = "East Asia"
    failover_priority = 1
  }
}
```
Then I update the ` enable_automatic_failover` to false. And it was applied successfully without error.

Please providing a more detailed reproducing steps if possible.

**Comment by rfojta** [comment]
Hello @yupwei68,

you need to create the account with non-standard prefixes, so anything random. Then you need to perform any change to the account, in our case we're changing list of IP addresses on the firewall, but it's non-relevant what change. Please enable trace log on the terraform

TF_LOG="trace"
TF_LOG_PATH="some_path/file.log"

That should trigger the error.

**Comment by yupwei68** [comment]
Hi @rfojta Sorry that from the support's log, I could not reproduce your error. Would you mind updating the latest Azurerm version and retry?
I've created a same cosmos account as you, and I applied it with success.

`{"kind":"MongoDB","location":"southeastasia","properties":{"consistencyPolicy":{"defaultConsistencyLevel":"Session","maxStalenessPrefix":100,"maxIntervalInSeconds":5},"locations":[{"failoverPriority":0,"locationName":"Southeast Asia"},{"failoverPriority":1,"locationName":"East Asia"}],"databaseAccountOfferType":"Standard","ipRules":[{"ipAddressOrRange":"52.139.237.121"},{"ipAddressOrRange":"52.139.235.113"}],"isVirtualNetworkFilterEnabled":false,"enableAutomaticFailover":true,"capabilities":[{"name":"EnableMongo"}],"virtualNetworkRules":[],"enableMultipleWriteLocations":true,"enableFreeTier":false},"tags":{}}`

**Comment by rfojta** [comment]
I've tried it with the latest version 2.33 and this is the error I got 

`
Error: Error updating CosmosDB Account "iws-test-eus2-uns-mongo36" properties (Resource Group "iws-test-eus2-uns"): Error creating/updating CosmosDB Account "iws-test-eus2-uns-mongo36" (Resource Group "iws-test-eus2-uns"): documentdb.DatabaseAccountsClient#CreateOrUpdate: Failure sending request: StatusCode=400 -- Original Error: Code="BadRequest" Message="Regions East US 2 already exists for database account iws-test-eus2-uns-mongo36.\r\nActivityId: 4c3fce4a-a332-43dc-a931-13b680a061f9, Microsoft.Azure.Documents.Common/2.11.0"
`

You need to use prefix in the `geo_location` block for example:
```
  geo_location {
    prefix             = "testing-east-asia-rep1"
    location          = "East Asia"
    failover_priority = 1
  }
```

You may need to add another region with the prefix to change the prefix. I have found that commenting out the line with "prefix" definition doesn't change a bit, which may lead to another bug.

Thank you.

**Comment by rfojta** [comment]
Hello, is there any update on reproducing the bug?

**Comment by rfojta** [comment]
@yupwei68 please respond

**Comment by yupwei68** [comment]
@rfojta Looks like our support colleague is contacting the service team for the ETA. Sorry that it's an upstream service problem. We'll need to wait for the response from the service team.

**Comment by rfojta** [comment]
Thank you @yupwei68, is there any info about what the upstream service is?

**Comment by gd-asharov** [comment]
Any updates here?

**Comment by dantdj** [comment]
Echoing the call for any updates!

Quick additional note - using v2.38 of the AzureRM provider fails every time, but using v2.25 fails the first time but appears to succeed after that.

**Comment by dantdj** [comment]
@yupwei68 - Please see the above comment. Are you able to comment on what upstream service we're waiting on, as well?

**Comment by yupwei68** [comment]
Hi, there are several issues here:
1. "prefix" in "geo_location" has been deprecated after azurerm 2.0.0, it's not supported to customize a location id any more after service version 2020-04-01.

2. The bug here is for the cosmosdb which has been created before azurerm 2.0.0, which has customized location id by setting "prefix" in "geo_location", could not be updated any more after azurerm 2.0.0. Please correct me if you're not the case. The service team has confirmed the problem and promise to fix it. I'm asking them for an ETA.

3. Another issue, if we change the other fields in cosmos account, the "geo_location" will be recreated. This will be fixed after we discuss it with service team.

Please correct me if I misunderstands the issues here.

**Comment by yupwei68** [comment]
The fix for the second problem, the update for the cosmosdb account with customized location id can not be applied, will be deployed by the end of this week. I'll keep following the updates.

**Comment by dantdj** [comment]
Thanks for the updates! Is that fix being deployed in Azure, or will it be a new version of the provider?

**Comment by yupwei68** [comment]
@dantdj It'll be deployed in the current version.

**Comment by dantdj** [comment]
@yupwei68 Thanks for confirming - will look more next week.

**Comment by yupwei68** [comment]
Hi, I've tested it. As the service team claims, the update for the cosmos account with customized location id could be applied. The bug has been fixed.

**Comment by favoretti** [comment]
Thanks for opening this issue! Since this issue seems to have been addressed in the latest versions of the provider (or a valid workaround was provided) - I'm going to close it. Please open a new updated bug report if this is still relevant. Thank you.
